consensus/src/chained_bft/liveness/pacemaker_timeout_manager.rs

```diff
@@ -109,38 +110,52 @@ impl PacemakerTimeoutManager {
     /// Returns the highest round PacemakerTimeoutCertificate from a map of author to
     /// timeout messages or None if there are not enough timeout messages available.
     /// A PacemakerTimeoutCertificate is made of the N highest timeout messages received where
-    /// N=timeout_quorum_size.  The round of PacemakerTimeoutCertificate is determined as
+    /// N=timeout_quorum_voting_power.  The round of PacemakerTimeoutCertificate is determined as
     /// the smallest of round of all messages used to generate this certificate.
     ///
-    /// For example, if timeout_certificate_quorum_size=3 and we received unique author timeouts
-    /// for rounds (1,2,3,4), then rounds (2,3,4) would form PacemakerTimeoutCertificate with
-    /// round=2.
+    /// For example, if timeout_certificate quorum_voting_power=3 and we received unique author
+    /// timeouts with equal voting power for rounds (1,2,3,4), then rounds (2,3,4) would form
+    /// PacemakerTimeoutCertificate with round=2.
     fn generate_timeout_certificate(
         author_to_received_timeouts: &HashMap<Author, PacemakerTimeout>,
-        quorum_size: usize,
+        validator_verifier: Arc<ValidatorVerifier>,
     ) -> Option<PacemakerTimeoutCertificate> {
-        if author_to_received_timeouts.values().len() < quorum_size {
-            return None;
+        // Sort the timeouts by round, highest to lowest and then, in order, aggregate
+        // enough voting power to assemble a PacemakerTimeoutCertificate if possible.
+        let mut received_timeouts: Vec<PacemakerTimeout> =
```

Comment:
you don't need clone it right? Vec<&PacemakerTimeout> should be enough given you already clone in line 135